### PR TITLE
Holocene: use extraData instead of nonce for eip-1559 parameters

### PR DESCRIPTION
**Pull request by roberto-bayardo**
**Description**

Use extraData for holocene EIP-1559 parameters so it will be propagated along with the execution payload.

**Additional context**

Corresponding spec update: https://github.com/ethereum-optimism/specs/pull/428

### CODE REVIEW

File: beacon/engine/types.go
Comment (tynes) [inline beacon/engine/types.go]: looks like a dead print statement

Comment (roberto-bayardo) [reply beacon/engine/types.go]: oops, yup leftover from debugging. removed.

**Comment by tynes** [comment]
Definitely most simple to just pass through the encoded params vs having them in a more human readable form. In theory, if there is an update to the serialization, we would be able to use hardfork branching to modify the way they are deserialized

**Review by tynes — APPROVED** [review]
Looks good to me

**Comment by roberto-bayardo** [comment]
I'm going to change this to use ExecutionPayloadEnvelope as discussed on the call.

**Comment by roberto-bayardo** [comment]
> I'm going to change this to use ExecutionPayloadEnvelope as discussed on the call.

Turned out to not be a great solution. Latest plan is to move from nonce to extradata for the params. extraData is already in the execution payload, so it simplifies things a lot.  Plus we probably want to use Nonce for sub-second block increments later on.

File: consensus/misc/eip1559/eip1559.go
Comment (protolambda) [inline consensus/misc/eip1559/eip1559.go]: See spec comment. It should check the hardfork time of the `parent` header if it's going to make assumptions on the `parent` header contents.
On the first block of the hardfork, with a pre-holocene parent-header, we could default to the old Canyon constants.

File: eth/catalyst/api.go
Comment (protolambda) [inline eth/catalyst/api.go]: `eip1559Params = bytes.Clone(payloadAttributes.EIP1559Params)`
may be cleaner than manual bytes length / copy

**Review by protolambda — APPROVED** [review]
Looks good. Can you update the `fork.yaml` to describe the changes to the files (also including some prior changes of e.g. introducing `EIP1559Params` to the payload attributes type)?

File: consensus/misc/eip1559/eip1559.go
Comment (protolambda) [reply consensus/misc/eip1559/eip1559.go]: I optimistically approved the PR, but this should be addressed before we can merge

Comment (roberto-bayardo) [reply consensus/misc/eip1559/eip1559.go]: done

**Comment by roberto-bayardo** [comment]
> Looks good. Can you update the `fork.yaml` to describe the changes to the files (also including some prior changes of e.g. introducing `EIP1559Params` to the payload attributes type)?

yes. ok to do that in another PR? Want to get this in to unblock the op-node reviews

File: eth/catalyst/api.go
Comment (roberto-bayardo) [reply eth/catalyst/api.go]: nice done

File: miner/payload_building_test.go
Comment (tynes) [inline miner/payload_building_test.go:130]: I don't notice in the spec any mention of the genesis block definition of the extra data if holocene is active at genesis. This is definitely a place where a bug is waiting. For the withdrawals root feature, we did specify that it needs to be explicitly empty in the genesis state for simplicity. We likely want initial values in the extradata if holocene is active at genesis so that the values can be used during execution. This could be a good detail to add to the spec

Comment (roberto-bayardo) [reply miner/payload_building_test.go:130]: It's implicitly required for genesis because we state it's required for all Holocene blocks.  I just updated the spec PR to call it out explicitly.

**Review by tynes — APPROVED** [review]

**Comment by sebastianst** [comment]
Created tag `v1.101411.1-rc.1` containing this change.

**Pull request merged**
